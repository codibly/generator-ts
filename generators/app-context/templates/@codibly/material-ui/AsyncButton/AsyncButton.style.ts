import styled from '@emotion/styled';
import { CircularProgress } from '@material-ui/core';
import { CircularProgressProps } from '@material-ui/core/CircularProgress';
import { ComponentType } from 'react';

export const ButtonProgress = styled(CircularProgress)({
  position: 'absolute'
}) as ComponentType<CircularProgressProps>;
