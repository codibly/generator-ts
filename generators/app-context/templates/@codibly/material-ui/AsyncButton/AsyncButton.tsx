import Button, { ButtonProps } from '@material-ui/core/Button';
import * as React from 'react';
import { FunctionComponent } from 'react';
import { ButtonProgress } from './AsyncButton.style';

export namespace AsyncButton {
  export type Props = ButtonProps & {
    pending?: boolean;
  };
}

export const AsyncButton: FunctionComponent<AsyncButton.Props> = ({
  disabled,
  pending,
  children,
  ...props
}) => (
  <Button disabled={disabled || pending} {...props}>
    {pending && <ButtonProgress size={24} />}
    {children}
  </Button>
);
