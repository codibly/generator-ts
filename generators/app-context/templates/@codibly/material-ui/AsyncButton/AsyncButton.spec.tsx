import * as React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { AsyncButton } from './AsyncButton';

describe('AsyncButton', () => {
  it('should render in default state', () => {
    const onClick = jest.fn();

    const { container, getByText } = render(
      <AsyncButton onClick={onClick}>Test button</AsyncButton>
    );

    const button = getByText('Test button');
    expect(button).toBeDefined();
    expect(onClick).not.toBeCalled();

    fireEvent.click(button);
    expect(onClick).toBeCalled();

    expect(container.firstChild).toMatchSnapshot();
  });

  it.each([[false, false, true], [false, true, false], [true, false, false], [true, true, false]])(
    'should render in pending: %p, disabled: %p state',
    (pending, disabled, clickable) => {
      const onClick = jest.fn();

      const { container, getByText } = render(
        <AsyncButton pending={pending} disabled={disabled} onClick={onClick}>
          Test button
        </AsyncButton>
      );

      const button = getByText('Test button');
      expect(button).toBeDefined();
      expect(onClick).not.toBeCalled();

      fireEvent.click(button);
      if (clickable) {
        expect(onClick).toBeCalled();
      } else {
        expect(onClick).not.toBeCalled();
      }

      expect(container.firstChild).toMatchSnapshot();
    }
  );
});
