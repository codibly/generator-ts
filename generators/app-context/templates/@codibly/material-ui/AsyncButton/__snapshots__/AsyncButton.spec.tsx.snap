// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncButton should render in default state 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text"
  tabindex="0"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Test button
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;

exports[`AsyncButton should render in pending: false, disabled: false state 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text"
  tabindex="0"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Test button
  </span>
  <span
    class="MuiTouchRipple-root"
  />
</button>
`;

exports[`AsyncButton should render in pending: false, disabled: true state 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled"
  disabled=""
  tabindex="-1"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    Test button
  </span>
</button>
`;

exports[`AsyncButton should render in pending: true, disabled: false state 1`] = `
.emotion-0 {
  position: absolute;
}

<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled"
  disabled=""
  tabindex="-1"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    <div
      class="MuiCircularProgress-root emotion-0 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style="width: 24px; height: 24px;"
    >
      <svg
        class="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </div>
    Test button
  </span>
</button>
`;

exports[`AsyncButton should render in pending: true, disabled: true state 1`] = `
.emotion-0 {
  position: absolute;
}

<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled Mui-disabled"
  disabled=""
  tabindex="-1"
  type="button"
>
  <span
    class="MuiButton-label"
  >
    <div
      class="MuiCircularProgress-root emotion-0 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style="width: 24px; height: 24px;"
    >
      <svg
        class="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </div>
    Test button
  </span>
</button>
`;
